### PR TITLE
Withdrawals processor

### PR DESCRIPTION
**Pull request by jframe**
## PR description
Process withdrawals when creating a block and when processing a block. Withdrawals processing is only enabled for the shanghai fork and beyond.

## Documentation

- [x] I thought about documentation and added the `doc-change-required` label to this PR if
    [updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

**Review by macfarla — COMMENTED** [review]
looks sensible to me

File: ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/AbstractBlockCreatorTest.java [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.units.bigints.UInt64;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class AbstractBlockCreatorTest {
```
Comment (siladu) [inline ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/AbstractBlockCreatorTest.java:61]: Nice! Surprised this didn't have a test before!

File: ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/PoWBlockCreatorTest.java [hunk 2]
```diff
@@ -52,9 +52,9 @@
 
 import com.google.common.collect.Lists;
 import org.apache.tuweni.bytes.Bytes;
-import org.junit.Test;
+import org.junit.jupiter.api.Test;
 
-public class PoWBlockCreatorTest {
+public class PoWBlockCreatorTest extends AbstractBlockCreatorTest {
```
Comment (siladu) [inline ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/PoWBlockCreatorTest.java:57]: If AbstractBlockCreatorTest isn't abstract, do we need to extend it here?

Comment (siladu) [inline ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/PoWBlockCreatorTest.java:57]: Do you think we need a MergeBlockCreatorTest as well?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/WithdrawalsProcessor.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  class DefaultWithdrawalsProcessor implements WithdrawalsProcessor {
+
+    @Override
+    public void processWithdrawals(
+        final List<Withdrawal> withdrawals, final WorldUpdater worldUpdater) {
+      for (final Withdrawal withdrawal : withdrawals) {
+        final EvmAccount account = worldUpdater.getOrCreate(withdrawal.getAddress());
+        account.getMutable().incrementBalance(withdrawal.getAmount());
+      }
+      worldUpdater.commit();
```
Comment (siladu) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/WithdrawalsProcessor.java:43]: Presumably it's possible for this to throw exceptions? Wonder what happens when it does.

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/mainnet/WithdrawalsProcessorTest.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                Wei.of(200)));
+    defaultWithdrawalsProcessor.processWithdrawals(withdrawals, updater);
+
+    assertThat(worldState.get(Address.fromHexString("0x1")).getBalance()).isEqualTo(Wei.of(101));
+    assertThat(worldState.get(Address.fromHexString("0x2")).getBalance()).isEqualTo(Wei.of(202));
+    assertThat(worldState.get(Address.fromHexString("0x3")).getBalance()).isEqualTo(Wei.of(3));
+  }
+
+  @Test
+  void defaultProcessor_shouldProcessWithdrawalsUpdatingEmptyAccountsBalance() {
```
Comment (siladu) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/mainnet/WithdrawalsProcessorTest.java:109]: good test, this was a bug that caught us out before!

File: ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/PoWBlockCreatorTest.java (hunk 2 above)
Comment (jframe) [reply ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/PoWBlockCreatorTest.java:57]: Hmm.. good point mostly wanted the name to match the test class as I put common logic in AbstractBlockCreator. Rename it to BaseBlockCreatorTest maybe?

Comment (jframe) [reply ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/PoWBlockCreatorTest.java:57]: We should have one, the only thing different about the MergeBlockCreator is the block header it creates, but it would be good to have a test for that. Do it this PR or another?

File: consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinatorTest.java [hunk 5]
```diff
@@ -213,7 +213,8 @@ public void coinbaseShouldMatchSuggestedFeeRecipient() {
             genesisState.getBlock().getHeader(),
             System.currentTimeMillis() / 1000,
             Bytes32.ZERO,
-            suggestedFeeRecipient);
+            suggestedFeeRecipient,
+            Optional.empty());
```
Comment (siladu) [inline consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinatorTest.java:217]: nit: could extract a variable to make this clearer it's emptyWithdrawals

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/WithdrawalsProcessor.java (hunk 3 above)
Comment (jframe) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/WithdrawalsProcessor.java:43]: It could throw a StorageException, this would be caught by the MainnetBlockValidator and logged. Do you think it would be worth adding some specific log errors for withdrawals?

File: ethereum/blockcreation/src/main/java/org/hyperledger/besu/ethereum/blockcreation/AbstractBlockCreator.java [hunk 6]
```diff
@@ -170,6 +174,13 @@ protected BlockCreationResult createBlock(
       final ProtocolSpec newProtocolSpec =
           protocolSchedule.getByBlockHeader(processableBlockHeader);
 
+      final WithdrawalsProcessor withdrawalsProcessor = newProtocolSpec.getWithdrawalsProcessor();
+      maybeWithdrawals.ifPresent(
+          withdrawals ->
+              withdrawalsProcessor.processWithdrawals(withdrawals, disposableWorldState.updater()));
+
+      throwIfStopped();
```
Comment (siladu) [inline ethereum/blockcreation/src/main/java/org/hyperledger/besu/ethereum/blockcreation/AbstractBlockCreator.java:182]: Interesting, I see this has moved to after withdrawals (compared to the prototype) - can you explain the significance please?

File: ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/PoWBlockCreatorTest.java (hunk 2 above)
Comment (siladu) [reply ethereum/blockcreation/src/test/java/org/hyperledger/besu/ethereum/blockcreation/PoWBlockCreatorTest.java:57]: Just thinking that PoW will rerun AbstractBlockCreatorTest's tests as well as the parent I think - is that the intention?

Potentially could make AbstractBlockCreatorTest abstract so they run for subclasses but not the parent

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/AbstractBlockProcessor.java [hunk 7]
```diff
@@ -134,6 +139,11 @@ public BlockProcessingResult processBlock(
       receipts.add(transactionReceipt);
     }
 
+    final WithdrawalsProcessor withdrawalsProcessor =
+        protocolSchedule.getByBlockHeader(blockHeader).getWithdrawalsProcessor();
+    maybeWithdrawals.ifPresent(
+        withdrawals -> withdrawalsProcessor.processWithdrawals(withdrawals, worldState.updater()));
```
Comment (siladu) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/AbstractBlockProcessor.java:145]: Just want to confirm my understanding: if we were pre-shanghai, but had withdrawals, we'd run the NoOpWithdrawalsProcessor and not error?

I think the api validation would prevent this, but wondering if we maybe log a warning in the NoOp just in case?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PrivacyBlockProcessor.java [hunk 8]
```diff
@@ -100,7 +102,13 @@ public BlockProcessingResult processBlock(
 
     final BlockProcessingResult result =
         blockProcessor.processBlock(
-            blockchain, worldState, blockHeader, transactions, ommers, metadataUpdater);
+            blockchain,
+            worldState,
+            blockHeader,
+            transactions,
+            ommers,
+            withdrawals,
```
Comment (siladu) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PrivacyBlockProcessor.java:110]: To play 'mainnet devil's advocate', what's the reasoning for changing PrivacyBlockProcessor as well?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/WithdrawalsProcessor.java (hunk 3 above)
Comment (siladu) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/WithdrawalsProcessor.java:43]: As long as we end up with an invalid block and handle that the right way, not sure if there's value in it being withdrawals specific

File: ethereum/blockcreation/src/main/java/org/hyperledger/besu/ethereum/blockcreation/AbstractBlockCreator.java (hunk 6 above)
Comment (jframe) [reply ethereum/blockcreation/src/main/java/org/hyperledger/besu/ethereum/blockcreation/AbstractBlockCreator.java:182]: There are multiple throwIfStopped calls. No significance in removing it,  think it makes sense to have it after as well, added that back in.

File: consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinatorTest.java (hunk 5 above)
Comment (jframe) [reply consensus/merge/src/test/java/org/hyperledger/besu/consensus/merge/blockcreation/MergeCoordinatorTest.java:217]: done

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/AbstractBlockProcessor.java (hunk 7 above)
Comment (jframe) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/AbstractBlockProcessor.java:145]: That's right, we would run the NoOpWithdrawalsProcessor.

Changing code around this area to return an Optional<WithdrawalProcessor> so there won't be a NoOpWithdrawalsProcessor at all. So will make that change first.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PrivacyBlockProcessor.java (hunk 8 above)
Comment (jframe) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/PrivacyBlockProcessor.java:110]: That it wouldn't compile without this change as I updated the full arg processBlock method to include Withdrawals. Would prefer not to touch this since it's going to be removed soon. But it's also a trivial change.

**Comment by jframe** [comment]
> Just thinking that PoW will rerun AbstractBlockCreatorTest's tests as well as the parent I think - is that the intention?
>
> Potentially could make AbstractBlockCreatorTest abstract so they run for subclasses but not the parent
Good point would be better to make this abstract

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/AbstractEngineForkchoiceUpdated.java [hunk 9]
```diff
@@ -162,7 +162,8 @@ public JsonRpcResponse syncResponse(final JsonRpcRequestContext requestContext)
                     newHead,
                     payloadAttributes.getTimestamp(),
                     payloadAttributes.getPrevRandao(),
-                    payloadAttributes.getSuggestedFeeRecipient()));
+                    payloadAttributes.getSuggestedFeeRecipient(),
+                    Optional.empty()));
```
Comment (siladu) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/AbstractEngineForkchoiceUpdated.java:166]: nit: Could pass in withdrawals at this point, or could keep in separate PR?

After moving the validation into this class, there's no refactor necessary, should just be a case of mapping the payloadAttributes.withdrawals -> List<Withdraw>...e.g. WithdrawalParameter.toDomain()

**Review by siladu — APPROVED** [review]

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/AbstractEngineForkchoiceUpdated.java (hunk 9 above)
Comment (jframe) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/AbstractEngineForkchoiceUpdated.java]: done

Comment (jframe) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/AbstractEngineForkchoiceUpdated.java]: done
